### ISSUE TITLE
Online Vocabulary Update Gives Bad Results With My Own Model

### ISSUE THREAD

**Issue by PrepareFunctionCall** [alphacep/vosk-api#788]
Hello, 

I'm new to Vosk and Kaldi and could use some assistance. 

When I run the test_simple.py and test_words.py with the small german default model - provided here: [Vosk Models](https://alphacephei.com/vosk/models) - everything is working as it should.

If I do it with my own model however, the result is really bad. I created the hclr and gr with the [mkgraph_lookahead.sh](https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/mkgraph_lookahead.sh). The model is also working perfectly fine when using the test_simple.py or test_microphone.py. However, once I use the functionality to adjust the vocabulary on the fly the performance plummets. The results are then empty strings or ukn - if the ukn token is included. 

The kaldi recognizer line looks like this:
```
rec = KaldiRecognizer(model, wf.getframerate(), '["licht an aus", "licht aus", "licht an", "an", "aus", "\<UNK>"]')
```
The log I get for building the new grammar looks like this:

```
LOG (VoskAPI:KaldiRecognizer():kaldi_recognizer.cc:63) ["licht an aus", "licht aus", "licht an", "an", "aus", "\<UNK>"]
LOG (VoskAPI:Estimate():language_model.cc:142) Estimating language model with ngram-order=2, discount=0.5
LOG (VoskAPI:OutputToFst():language_model.cc:209) Created language model with 5 states and 11 arcs.
```

One thing of note is that the unknown token is different in the lm I use, it's not working with [unk]. Might a difference in token names/mapping be the problem?

The first two and last three words/tokens in words.txt look like this:
```
<eps> 0
<UNK> 1
.
.
#0 659356
<s> 659357
</s> 659358
```

Another observation I made, is that the mkgraph_lookahead.sh fails, when trying to use my ARPA file with:

```
apply_map.pl: warning! missing key 0 in /opt/speech_environment/model_output/relabel
apply_map.pl: warning! missing key 659358 in /opt/speech_environment/model_output/relabel
FATAL: NGramInput: Have a backoff cost with no state ID!
ERROR: FstHeader::Read: Bad FST header: standard input
ERROR: FstHeader::Read: Bad FST header: standard input
```

However, if I don't use the ARPA file to build the lookahead graph and instead use the G.fst created from the same ARPA file with Kaldis [format_lm.sh](https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/format_lm.sh) the HCLr and Gr are build without issue and the model also works with mic/wave file. Just the online vocabulary updating gives bad results. 

If you have any idea on what the problem is and how to solve it, I'd be grateful.

**Comment by nshmyrev** [comment]
> If you have any idea on what the problem is and how to solve it, I'd be grateful.

You probably used original words.txt in the model while you need to use the LM with relabelled symbols (they look like have integer ids in random order).  You can print the values of the dynamically constructed graph inside the recognizer and compare with required word ids.

`<UNK>` for unknown is ok if you have it in the dictionary instead of "[unk]"

**Comment by PrepareFunctionCall** [comment]
Thank you for your answer! 

That is indeed the core of the issue. But, I feel that something went really wrong somewhere. 

The words.txt before the mkgraph_lookahead.sh looks like this (it is ordered as you mentioned): 

![image](https://user-images.githubusercontent.com/90621747/146326213-c9b1f551-7e54-4cfd-8120-e793b4cfa768.png)

The output words.txt after mkgraph_lookahead.sh, however, looks like this:

![image](https://user-images.githubusercontent.com/90621747/146326466-58c6ba60-603e-4493-b23e-13cc6b9063c1.png)


It did not get relabeled. So I did the relabeling myself, using the relabel file present in the folder, to get a words.txt that looks like this: 

![image](https://user-images.githubusercontent.com/90621747/146326666-72c63d8c-f599-4909-b012-fca691246524.png)

And now comes the odd part. When I use the test_simple.py it basically gives me random words. 
Probably the wrong mappings. 
But using the test_words.py now works like a charm. So after the rebuilding, it's doing exactly what it should be doing.

**Comment by nshmyrev** [comment]
> When I use the test_simple.py it basically gives me random words. Probably the wrong mappings.

Probably your G.fst is not properly relabelled as well. Not sure how you came up with it.

**Comment by PrepareFunctionCall** [comment]
Okay, yes that is the issue. So I tried using mkgraph_lookahead.sh with the lm.gz file, but that failed with:

apply_map.pl: warning! missing key 0 in /opt/speech_environment/model_output/relabel
apply_map.pl: warning! missing key 659358 in /opt/speech_environment/model_output/relabel
FATAL: NGramInput: Have a backoff cost with no state ID!
ERROR: FstHeader::Read: Bad FST header: standard input
ERROR: FstHeader::Read: Bad FST header: standard input 

So I created the G.fst with [format_lm.sh](https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/format_lm.sh) and then the building of the lookahead graph worked without feeding the lm.gz. But since the G.fst is now already present, it didn't rebuild/relable it - I assume.

So I need to fix the errors I get when feeding the lm.gz file.
